modules/nlu/src/backend/predict-pipeline.ts

```diff
@@ -80,8 +85,37 @@ async function DetectLanguage(
     detectedLanguage = NA_LANG
   }
 
-  const usedLanguage =
-    detectedLanguage !== NA_LANG && elected.value > threshold ? detectedLanguage : input.defaultLanguage
+  // if ML-based language identifier didn't find a match
+  // we proceed with a custom vocabulary matching algorithm
+  // ie. the % of the sentence comprised of tokens in the training vocabulary
+  if (detectedLanguage === NA_LANG) {
+    try {
+      const match = _.chain(supportedLanguages)
+        .map(lang => ({
+          lang,
+          sentence: input.sentence.toLowerCase(),
+          tokens: _.orderBy(Object.keys(predictorsByLang[lang].vocabVectors), 'length', 'desc')
+        }))
+        .map(({ lang, sentence, tokens }) => {
+          for (const token of tokens) {
```

Comment:
this could be done in a single regex

Reply:
what do you mean?  appending all tokens together?  regex length evaluation is exponential with length so wouldn't do that

Reply:
right